### PR TITLE
✨Quicksettings Integration for android 

### PR DESCRIPTION
**Pull request by strseb**
Small thing i wrote while waiting for taskcluster jobs 😅
Adds a tile-service that can interact with the vpn-Service.
The tile service is registered with the OS, thus adding a quick-settings option to control "on/off" on the VPN. 

https://user-images.githubusercontent.com/9611612/155134283-5b0507e0-047b-4602-bb94-b802d0357405.mp4

### CODE REVIEW

**Comment by strseb** [comment]
@albionx - I just found this PR - it bit rotted a bit already. 
In general what do you think, should we close or try to have it ? :)

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        super.onClick()
+    }
+
+    private fun updateTile() {
+        qsTile.apply {
+            when (mState) {
+                State.Error -> {
+                    state = Tile.STATE_UNAVAILABLE
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                        stateDescription = "Error CommuniCating with service"
```
Comment (brizental) [inline android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt:106]:
```suggestion
                        stateDescription = "Error communicating with service"
```

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                    state = Tile.STATE_INACTIVE
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                        subtitle = "Disconnected"
+                    }
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                        stateDescription = "Not Connected with VPN"
+                    }
+                }
+            }
+        }
```
Comment (brizental) [inline android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt:134]: That is some annoying version checking... Does it error without it? Android should just ignore the unknown stuff...

**Review by brizental — CHANGES_REQUESTED** [review]
This is a great new feature, in my opinion. :100: 

iOS has a similar thing, do we have a ticket to implement it there? If not, would you mind filing a follow up?

The code looks good to me, I left small nits. The main issue I'd say is the addressing of error cases. Maybe something we can bring UX on to look at and address as a follow up too.

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt (hunk 2 above)
Comment (strseb) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt:133]: Removed some redudancy :) 
Not checking not an option as ```stateDescription = "Not Connected with VPN"``` will be translated into ```qsTile.setDescription(<>)``` and calling nonexistent functions is nothing the jvm likes >:c

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNServiceBinder.kt [hunk 3]
```diff
@@ -88,6 +90,16 @@ class VPNServiceBinder(service: VPNService) : Binder() {
...
+            ACTIONS.reactivate -> {
+                // [data] is empty
+                // Activate the tunnel with the last config
+                try {
+                    this.mService.reconnect()
+                } catch (e: Exception) {
+                    Log.e(tag, "An Error occurred while enabling the VPN: ${e.localizedMessage}")
+                }
+                return true
+            }
```
[the remark this answers is not in the document]
Comment (strseb) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNServiceBinder.kt:103]: Good callout! I've added the option for the tile-server to accept activation errors, so it should move into a "deactivated & disabled " state.

**Comment by strseb** [comment]
> iOS has a similar thing, do we have a ticket to implement it there? If not, would you mind filing a follow up? 

We have had an issue about that, but i think the verdict was that there is no option for iOS to hook us into that surface.

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt (hunk 2 above)
Comment (brizental) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt:135]: Boo JVM!

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNServiceBinder.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (brizental) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNServiceBinder.kt:103]: Would it be better to actually move into an error state so the user knows why it's not turning on even as the yclick to do so? No need to do that on this PR if it makes sense.

**Review by brizental — APPROVED** [review]

**Pull request merged**